kernel/kobject.c

```diff
@@ -226,23 +227,44 @@ int kobject_lookup( struct kobject *kobject, const char *name, struct kobject **
 	return 0;
 }
 
-int kobject_copy( struct kobject *ksrc, struct kobject *kdst )
+int kobject_copy( struct kobject *ksrc, struct kobject **kdst )
 {
-	struct fs_dirent *src, *dst;
-
-	if(ksrc->type==KOBJECT_DIR) {
-		src = ksrc->data.dir;
-	} else {
-		return KERROR_NOT_A_FILE;
-	}
-
-	if(kdst->type==KOBJECT_DIR) {
-		dst = kdst->data.dir;
-	} else {
-		return KERROR_NOT_A_DIRECTORY;
+	/*
+		Shallow copy the state of ksrc to kdst and call addref for underlying
+		type.
+	*/
+	*kdst = kobject_init();
+	(*kdst)->refcount 			= 1;
+	(*kdst)->data 			= ksrc->data;
+	(*kdst)->type 			= ksrc->type;
+	(*kdst)->offset 		= ksrc->offset;
+	(*kdst)->tag 				= ksrc->tag;
+	(*kdst)->blocking 	= ksrc->blocking;
+
+	switch (ksrc->type) {
+	case KOBJECT_GRAPHICS:
+		graphics_addref(ksrc->data.graphics);
+		break;
+	case KOBJECT_CONSOLE:
+		console_addref(ksrc->data.console);
+		break;
+	case KOBJECT_FILE:
+		fs_dirent_addref(ksrc->data.file);
+		break;
+	case KOBJECT_DIR:
+		fs_dirent_addref(ksrc->data.dir);
+		break;
+	case KOBJECT_DEVICE:
+		device_addref(ksrc->data.device);
+		break;
+	case KOBJECT_PIPE:
+		pipe_addref(ksrc->data.pipe);
+		break;
+	default:
```

Comment:
As a style point, I would *not* have a default case here.
The compiler can check whether we used all of the switch cases here.
If a new one gets added (and no default is provided) then the compiler will tell us.